File: backend/app/models/rice_config.py
import re
from typing import Annotated, Literal, Union
from pydantic import BaseModel, Field, field_validator


# ─── Helpers ──────────────────────────────────────────────────────────────────

_HEX_RE = re.compile(r"^#[0-9a-fA-F]{6}$")


def _validate_hex(v: str) -> str:
    if not _HEX_RE.match(v):
        raise ValueError(f"'{v}' is not a valid hex color (#rrggbb)")
    return v.lower()


HexColor = Annotated[str, Field(pattern=r"^#[0-9a-fA-F]{6}$")]


# ─── Color Palette ────────────────────────────────────────────────────────────

class ColorPalette(BaseModel):
    accent: HexColor
    background: HexColor
    foreground: HexColor
    border: HexColor
    surface: HexColor

    @field_validator("accent", "background", "foreground", "border", "surface", mode="before")
    @classmethod
    def normalize_hex(cls, v: str) -> str:
        return _validate_hex(v)


# ─── Font Config ──────────────────────────────────────────────────────────────

class FontConfig(BaseModel):
    heading_family: str = Field(min_length=1, max_length=100)
    body_family: str = Field(min_length=1, max_length=100)
    mono_family: str = Field(min_length=1, max_length=100)
    size: int = Field(ge=6, le=72)


# ─── WM Config ────────────────────────────────────────────────────────────────

class HyprlandConfig(BaseModel):
    kind: Literal["hyprland"]
    gaps_in: int = Field(ge=0, le=100)
    gaps_out: int = Field(ge=0, le=100)
    border_size: int = Field(ge=0, le=20)
    rounding: int = Field(ge=0, le=50)
    active_opacity: float = Field(ge=0.0, le=1.0)
    inactive_opacity: float = Field(ge=0.0, le=1.0)
    blur: bool
    blur_size: int = Field(ge=0, le=20)
    animations: bool
    shadow: bool
    wallpaper_image: str | None = None
    wallpaper_tool: Literal["hyprpaper", "swww", "swaybg"]


WMConfig = HyprlandConfig


# ─── Bar Config ───────────────────────────────────────────────────────────────

WaybarModule = Literal[
    "hyprland/workspaces",
    "hyprland/window",
    "hyprland/taskbar",
    "hyprland/submap",
    "clock",
    "custom/date",
    "custom/separator",
    "cpu",
    "memory",
    "disk",
    "temperature",
    "network",
    "bluetooth",
    "battery",
    "pulseaudio",
    "tray",
    "custom/power",
]


class WaybarConfig(BaseModel):
    kind: Literal["waybar"]
    position: Literal["top", "bottom"]
    height: int = Field(ge=16, le=80)
    spacing: int = Field(ge=0, le=40)
    modules_left: list[WaybarModule]
    modules_center: list[WaybarModule]
    modules_right: list[WaybarModule]
    show_icons: bool
    show_labels: bool


BarConfig = WaybarConfig


# ─── Terminal Config ──────────────────────────────────────────────────────────

class KittyConfig(BaseModel):
    kind: Literal["kitty"]
    font_family: str = Field(min_length=1, max_length=100)
    font_size: int = Field(ge=6, le=72)
    padding_x: int = Field(ge=0, le=100)
    padding_y: int = Field(ge=0, le=100)
    cursor_shape: Literal["block", "beam", "underline"]
    cursor_blink: bool


TerminalConfig = KittyConfig


# ─── App Launcher Config ──────────────────────────────────────────────────────

class RofiConfig(BaseModel):
    kind: Literal["rofi"]
    mode: Literal["launcher", "dmenu", "drun", "window"]
    width: int = Field(ge=200, le=2000)
    lines: int = Field(ge=1, le=50)
    position: Literal["left", "center", "right"]
    show_icons: bool


class WofiConfig(BaseModel):
    kind: Literal["wofi"]
    width: int = Field(ge=200, le=2000)
    height: int = Field(ge=100, le=2000)
    show_icons: bool


AppLauncherConfig = Annotated[
    Union[RofiConfig, WofiConfig],
    Field(discriminator="kind"),
]


# ─── Lock Screen Config ───────────────────────────────────────────────────────

class HyprlockConfig(BaseModel):
    kind: Literal["hyprlock"]
    theme: str = Field(min_length=1, max_length=100)
    background_color: HexColor
    background_image: str | None = None
    blur: bool
    blur_size: int = Field(ge=0, le=20)
    clock: bool
    date_format: str = Field(min_length=1, max_length=50)
    layout: Literal["center", "left", "right"]
    layout_y: Literal["top", "center", "bottom"]

    @field_validator("background_color", mode="before")
    @classmethod
    def normalize_hex(cls, v: str) -> str:
        return _validate_hex(v)


class SwaylockConfig(BaseModel):
    kind: Literal["swaylock"]
    theme: str = Field(min_length=1, max_length=100)
    color: HexColor
    background_image: str | None = None
    blur: bool
    clock: bool
    layout: Literal["center", "left", "right"]
    layout_y: Literal["top", "center", "bottom"]

    @field_validator("color", mode="before")
    @classmethod
    def normalize_hex(cls, v: str) -> str:
        return _validate_hex(v)


class SddmConfig(BaseModel):
    kind: Literal["sddm"]
    theme: str = Field(min_length=1, max_length=100)
    background_color: HexColor
    background_image: str | None = None
    font: str = Field(min_length=1, max_length=100)
    show_logo: bool
    blur: bool
    blur_size: int = Field(ge=0, le=20)
    layout: Literal["center", "left", "right"]
    layout_y: Literal["top", "center", "bottom"]

    @field_validator("background_color", mode="before")
    @classmethod
    def normalize_hex(cls, v: str) -> str:
        return _validate_hex(v)


LockScreenConfig = Annotated[
    Union[HyprlockConfig, SwaylockConfig, SddmConfig],
    Field(discriminator="kind"),
]


# ─── Root Config ──────────────────────────────────────────────────────────────

class RiceConfig(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    colors: ColorPalette
    font: FontConfig
    wm: WMConfig
    bar: BarConfig
    terminal: TerminalConfig
    launcher: AppLauncherConfig
    lockscreen: LockScreenConfig


# ─── Request / Response ───────────────────────────────────────────────────────

class GenerateRequest(BaseModel):
    rice_id: str = Field(min_length=1)
    user_id: str = Field(min_length=1)
    config: RiceConfig

File: frontend/src/types/rice-config.ts
// ─── Color Palette ────────────────────────────────────────────────────────────

export type ColorPalette = {
  accent: string;
  background: string;
  foreground: string;
  border: string;
  surface: string;
};

// ─── Font Config ──────────────────────────────────────────────────────────────

export type FontConfig = {
  heading_family: string;
  body_family: string;
  mono_family: string;
  size: number; // base size in px
};

// ─── WM Config ────────────────────────────────────────────────────────────────

export type WallpaperTool = "hyprpaper" | "swww" | "swaybg";

export type HyprlandConfig = {
  kind: "hyprland";
  gaps_in: number;
  gaps_out: number;
  border_size: number;
  rounding: number;
  active_opacity: number;
  inactive_opacity: number;
  blur: boolean;
  blur_size: number;
  animations: boolean;
  shadow: boolean;
  wallpaper_image?: string;
  wallpaper_tool: WallpaperTool;
};

export type WMConfig = HyprlandConfig;
// future: | SwayConfig | I3Config

// ─── Bar Config ───────────────────────────────────────────────────────────────

export type WaybarModule =
  | "hyprland/workspaces"
  | "hyprland/window"
  | "hyprland/taskbar"
  | "hyprland/submap"
  | "clock"
  | "custom/date"
  | "custom/separator"
  | "cpu"
  | "memory"
  | "disk"
  | "temperature"
  | "network"
  | "bluetooth"
  | "battery"
  | "pulseaudio"
  | "tray"
  | "custom/power";

export type WaybarConfig = {
  kind: "waybar";
  position: "top" | "bottom";
  height: number;
  spacing: number;
  modules_left: WaybarModule[];
  modules_center: WaybarModule[];
  modules_right: WaybarModule[];
  show_icons: boolean;
  show_labels: boolean;
};

export type BarConfig = WaybarConfig;
// future: | PolybarConfig

// ─── Terminal Config ──────────────────────────────────────────────────────────

export type CursorShape = "block" | "beam" | "underline";

export type KittyConfig = {
  kind: "kitty";
  font_family: string;
  font_size: number;
  padding_x: number;
  padding_y: number;
  cursor_shape: CursorShape;
  cursor_blink: boolean;
};

export type TerminalConfig = KittyConfig;
// future: | AlacrittyConfig | FootConfig

// ─── App Launcher Config ──────────────────────────────────────────────────────

export type RofiMode = "launcher" | "dmenu" | "drun" | "window";

export type RofiConfig = {
  kind: "rofi";
  mode: RofiMode;
  width: number;
  lines: number;
  position: "left" | "center" | "right";
  show_icons: boolean;
};

export type WofiConfig = {
  kind: "wofi";
  width: number;
  height: number;
  show_icons: boolean;
};

export type AppLauncherConfig = RofiConfig | WofiConfig;

// ─── Lock Screen Config ───────────────────────────────────────────────────────

export type LockLayout = "center" | "left" | "right";
export type LockLayoutY = "top" | "center" | "bottom";

export type HyprlockConfig = {
  kind: "hyprlock";
  theme: string;
  background_color: string;
  background_image?: string;
  blur: boolean;
  blur_size: number;
  clock: boolean;
  date_format: string;
  layout: LockLayout;
  layout_y: LockLayoutY;
};

export type SwaylockConfig = {
  kind: "swaylock";
  theme: string;
  color: string;
  background_image?: string;
  blur: boolean;
  clock: boolean;
  layout: LockLayout;
  layout_y: LockLayoutY;
};

export type SddmConfig = {
  kind: "sddm";
  theme: string;
  background_color: string;
  background_image?: string;
  font: string;
  show_logo: boolean;
  blur: boolean;
  blur_size: number;
  layout: LockLayout;
  layout_y: LockLayoutY;
};

export type LockScreenConfig = HyprlockConfig | SwaylockConfig | SddmConfig;

// ─── Root Config ──────────────────────────────────────────────────────────────

export type RiceConfig = {
  name: string;
  colors: ColorPalette;
  font: FontConfig;
  wm: WMConfig;
  bar: BarConfig;
  terminal: TerminalConfig;
  launcher: AppLauncherConfig;
  lockscreen: LockScreenConfig;
};

// ─── Kind Helpers ─────────────────────────────────────────────────────────────

export type WMKind = WMConfig["kind"];
export type BarKind = BarConfig["kind"];
export type TermKind = TerminalConfig["kind"];
export type LauncherKind = AppLauncherConfig["kind"];
export type LockKind = LockScreenConfig["kind"];

// ─── Constants ────────────────────────────────────────────────────────────────

export const SANS_FONTS = [
  "Geist Sans",
  "Inter",
  "IBM Plex Sans",
  "Noto Sans",
  "Roboto",
  "Ubuntu",
  "Cantarell",
] as const;

export const MONO_FONTS = [
  "JetBrains Mono",
  "Fira Code",
  "Cascadia Code",
  "Hack Nerd Font",
  "Iosevka Nerd Font",
  "Source Code Pro",
  "Inconsolata",
  "IBM Plex Mono",
] as const;

export const WAYBAR_MODULES: WaybarModule[] = [
  "hyprland/workspaces",
  "hyprland/window",
  "hyprland/taskbar",
  "hyprland/submap",
  "clock",
  "custom/date",
  "custom/separator",
  "cpu",
  "memory",
  "disk",
  "temperature",
  "network",
  "bluetooth",
  "battery",
  "pulseaudio",
  "tray",
  "custom/power",
];

